lib/src/revset.rs

```diff
@@ -437,64 +438,162 @@ impl RevsetExpression {
         Rc::new(RevsetExpression::Difference(self.clone(), other.clone()))
     }
 
+    /// Commits that are in `self` but not in any of the `others`.
+    pub fn minus_all(
+        self: &Rc<RevsetExpression>,
+        others: &[Rc<RevsetExpression>],
+    ) -> Rc<RevsetExpression> {
+        if others.is_empty() {
+            self.clone()
+        } else {
+            RevsetExpression::minus(self, &RevsetExpression::union_all(others))
+        }
+    }
+
     /// Internal helper for matching on signature fields.
     fn signature_field(
-        predicate: fn(SignatureField, StringPattern) -> RevsetFilterPredicate,
+        raw_predicate: fn(SignatureField, StringPattern) -> RevsetFilterPredicate,
         field: SignatureField,
         pattern: StringPattern,
+        mailmap: &Mailmap,
     ) -> Rc<RevsetExpression> {
-        RevsetExpression::filter(predicate(field, pattern))
+        // Matching against a canonical signature field is the same as matching against
+        // the corresponding raw signature field except:
+        //
+        // * field values that may match but are mapped to a value that does not should
+        //   be excluded; and
+        //
+        // * field values that may not match but are mapped to a value that does should
+        //   be included.
+        //
+        // We build the expression `(raw_field(pattern) - excludes) | includes`, where:
+        //
+        //     excludes = union_all {
+        //         entry : entry ∈ mailmap,
+        //         (old field absent | old field matches) &
+        //         (new field present & new field doesn’t match)
+        //     }
+        //
+        //     includes = union_all {
+        //         entry : entry ∈ mailmap,
+        //         (old field absent | old field doesn’t match) &
+        //         (new field present & new field matches)
+        //     }
+        //
+        // When matching raw signatures, the `Mailmap` is empty and this reduces to
+        // `raw_field(pattern)`.
```

Comment:
nit: it's probably better to write unit tests for this query transformation function.